engine/src/main/java/org/hibernate/search/engine/spi/DocumentBuilderIndexedEntity.java

```diff
@@ -803,12 +809,46 @@ else if ( numericValue instanceof Float ) {
 					}
 				}
 				else {
-					document.add( new SortedDocValuesField( sortField.getFieldName(), new BytesRef( field.stringValue() ) ) );
+					String value = field.stringValue();
+					AnalyzerReference analyzerReference = fieldMetaData.getAnalyzerReference();
+					if ( fieldMetaData.getIndex().isAnalyzed() && analyzerReference.is( LuceneAnalyzerReference.class ) ) {
+						// Analysis is skipped altogether when the analyzer is remote. It's up to the backend to handle it.
+						Analyzer analyzer = analyzerReference.unwrap( LuceneAnalyzerReference.class ).getAnalyzer();
+						value = analyzeSortFieldValue( sortField.getFieldName(), field.stringValue(), analyzer );
+					}
+					if ( value != null ) {
+						document.add( new SortedDocValuesField( sortField.getFieldName(), new BytesRef( value ) ) );
+					}
+				}
+			}
+		}
+	}
+
+	/*
+	 * Necessary because Lucene doesn't automatically analyze SortedDocValuesFields.
+	 * See https://hibernate.atlassian.net/browse/HSEARCH-2376
+	 */
+	private String analyzeSortFieldValue(String fieldName, String fieldValue, Analyzer analyzer) {
+		try {
+			Iterator<String> analyzedTokens = AnalyzerUtils.tokenizedTermValues( analyzer, fieldName, fieldValue ).iterator();
+
+			if ( analyzedTokens.hasNext() ) {
```

Comment:
I wouldn't have created an iterator and would have worked on the list here but I suppose it's a matter of taste :).